src/algo/tred.rs

```diff
@@ -0,0 +1,129 @@
+//! Compute the transitive reduction and closure of a directed acyclic graph
+//!
+//! ## Transitive reduction and closure
+//! The *transitive closure* of a graph **G = (V, E)** is the graph **Gc = (V, Ec)**
+//! such that **(i, j)** belongs to **Ec** if and only if there is a path connecting
+//! **i** to **j** in **G**. The *transitive reduction* of **G** is the graph **Gr
+//! = (V, Er)** such that **Er** is minimal wrt. inclusion in **E** and the transitive
+//! closure of **Gr** is the same as that of **G**.
+//! The transitive reduction is well-defined for acyclic graphs only.
+
+use crate::Direction;
+use crate::adj::List;
+use crate::data::{NodeIndexMap, NodeIndexMappable};
+use fixedbitset::FixedBitSet;
+use crate::graph::IndexType;
+use crate::visit::{GraphBase, IntoNeighbors, IntoNeighborsDirected, NodeCount};
+
+/// Creates a representation of the same graph respecting topological order for use in [`dag_transitive_reduction_closure`].
+///
+/// `toposort` must be a topological order on the node indices of `g` (for example obtained
+/// from [`toposort`]).
+///
+/// [`toposort`]: crate::algo::toposort
+///
+/// The resulting graph is the same as `g` with the following differences:
+/// * Node and edge weights are stripped,
+/// * Node indices are replaced by the corresponding rank in `toposort`,
+/// * Iterating on the neighbors of a node respects topological order.
+///
+/// Runtime: **O(|V| + |E|)**.
+///
+/// Space complexity: **O(|V| + |E|)**.
+pub fn dag_to_toposorted_adjacency_list<G, Ix: IndexType>(
+    g: G,
+    toposort: &[G::NodeId],
+) -> List<(), Ix>
```

Comment:
I believe a type alias for `List<(), Ix>` could make things a bit clearer. Maybe you could have other versions of these functions taking a node/edge mapping function, with the current functions passing `|_| ()` for both.

Reply:
I added `UnweighedList` as an alias for `List<()>`.